analyzers/src/SonarAnalyzer.Common/Rules/CheckFileLicenseBase.cs

```diff
@@ -18,31 +18,61 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-using Microsoft.CodeAnalysis;
-using SonarAnalyzer.Common;
-using SonarAnalyzer.Helpers;
 using System;
 using System.Collections.Immutable;
+using System.Linq;
 using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
 
 namespace SonarAnalyzer.Rules
 {
     public abstract class CheckFileLicenseBase : ParameterLoadingDiagnosticAnalyzer
     {
         internal const string DiagnosticId = "S1451";
-        protected const string MessageFormat = "Add or update the header of this file.";
-
         internal const string HeaderFormatRuleParameterKey = "headerFormat";
         internal const string HeaderFormatPropertyKey = nameof(HeaderFormat);
-        public abstract string HeaderFormat { get; set; }
-
         internal const string IsRegularExpressionRuleParameterKey = "isRegularExpression";
         internal const string IsRegularExpressionPropertyKey = nameof(IsRegularExpression);
         internal const string IsRegularExpressionDefaultValue = "false";
-        [RuleParameter(IsRegularExpressionRuleParameterKey, PropertyType.Boolean,
-            "Whether the headerFormat is a regular expression.", IsRegularExpressionDefaultValue)]
+        protected const string MessageFormat = "Add or update the header of this file.";
+
+        private readonly DiagnosticDescriptor rule;
+
+        protected abstract ILanguageFacade Language { get; }
+        public abstract string HeaderFormat { get; set; }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        [RuleParameter(IsRegularExpressionRuleParameterKey, PropertyType.Boolean, "Whether the headerFormat is a regular expression.", IsRegularExpressionDefaultValue)]
         public bool IsRegularExpression { get; set; } = bool.Parse(IsRegularExpressionDefaultValue);
 
+        protected CheckFileLicenseBase() =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
+
+        protected override void Initialize(ParameterLoadingAnalysisContext context) =>
+            context.RegisterSyntaxTreeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
+                {
+                    if (HeaderFormat == null)
+                    {
+                        return;
+                    }
+
+                    if (IsRegularExpression && !IsRegexPatternValid(HeaderFormat))
+                    {
+                        throw new InvalidOperationException($"Invalid regular expression: {HeaderFormat}");
```

Comment:
Do we really want to throw an exception here? If I understand the behavior right this will raise AD001 for each project file if the regex pattern is not valid.

Reply:
I don't remember any community problem with this so far. 
It's not SonarWay.
There's no validation on SQ/SC side for the regex.
If you want this blocker rule, you should have it configured it correctly instead of silent FNs everywhere that are hard to diagnose. 

So I'd say it's noisy, but reasonable in this case.